# Part 1: split

import file

In [109]:
import os
import pandas as pd
from pprint import pprint

# Specify the folder path
folder_path = './all data.csv'

df = pd.read_csv(folder_path)

filter data to have companies

In [110]:
import pandas as pd
from io import StringIO

# Extract the information from 'airport' to the first comma
df['departure airport names'] = df['departure'].apply(lambda x: x.split(',')[0][1:-1])
df['arrival airport names'] = df['arrival'].apply(lambda x: x.split(',')[0][1:-1])

# Display the updated DataFrame
print(df[['departure', 'arrival']])

# Create a new DataFrame with only 'departure' and 'arrival' columns
airpotnames_df = df[['departure airport names', 'arrival airport names']]

# Save the new DataFrame to a new CSV file
airpotnames_df.to_csv('airport_names.csv', index=False)

                                              departure  \
0     {'airport': 'Oenpelli', 'timezone': 'Australia...   
1     {'airport': 'Sydney Bankstown', 'timezone': 'A...   
2     {'airport': 'Haneda Airport', 'timezone': 'Asi...   
3     {'airport': 'Taiwan Taoyuan International (Chi...   
4     {'airport': 'Ninoy Aquino International', 'tim...   
...                                                 ...   
1594  {'airport': 'Singapore Changi', 'timezone': 'A...   
1595  {'airport': 'Auckland International', 'timezon...   
1596  {'airport': 'Auckland International', 'timezon...   
1597  {'airport': 'Melbourne - Tullamarine Airport',...   
1598  {'airport': 'Melbourne - Tullamarine Airport',...   

                                                arrival  
0     {'airport': 'Maningrida', 'timezone': 'Austral...  
1     {'airport': 'Williamtown', 'timezone': 'Austra...  
2     {'airport': 'Hong Kong International', 'timezo...  
3     {'airport': 'Singapore Changi', 'timezone': 'A...  
4

filter data to have delays

In [111]:
# Extract the information starting from 'delay': all the way to the first comma for the 'departure' column
df['delay in departure'] = df['departure'].apply(lambda x: x.split('delay\': ')[1].split(',')[0])
df['delay in arrival'] = df['arrival'].apply(lambda x: x.split('delay\': ')[1].split(',')[0])

delay_df = df[['delay in departure', 'delay in arrival']]
delay_df.to_csv('airport_delays.csv', index=False)

filter data to airway companies

In [112]:
df['airline names'] = df['airline'].apply(lambda x: x.split('name\': ')[1].split(',')[0])

airwaynames_df = df[['airline names']]
airwaynames_df.to_csv('airway_names.csv', index=False)

regions

In [113]:
df['departure region'] = df['departure'].apply(lambda x: x.split('timezone\': ')[1].split(',')[0])
df['arrival region'] = df['arrival'].apply(lambda x: x.split('timezone\': ')[1].split(',')[0])

region_df = df[['departure region', 'arrival region']]
region_df.to_csv('region.csv', index=False)

# Part 2: Analyze

### 1: delay data overall

In [123]:
# convert str to numeric
df['delay in departure'] = pd.to_numeric(df['delay in departure'], errors='coerce')
#avergae delay in departure:
print(df['delay in departure'].mean())

17.309114927344783


In [124]:
# convert str to numeric
df['delay in arrival'] = pd.to_numeric(df['delay in arrival'], errors='coerce')
#avergae delay in departure:
print(df['delay in arrival'].mean())

6.704444444444444


### 2: delay data with regions

2.1 departures

In [ ]:
df_departure_region =df[['delay in departure',  'departure region']]
df_departure_region

In [ ]:
mask_asia = df_departure_region['departure region'].str.contains('Asia', na=False)
df_asia_departures = df_departure_region[mask_asia]
df_asia_departures['delay in departure'].mean()


In [ ]:
mask_aus = df_departure_region['departure region'].str.contains('Australia', na=False)
df_aus_departures = df_departure_region[mask_aus]
df_aus_departures['delay in departure'].mean()

In [ ]:
mask_pac = df_departure_region['departure region'].str.contains('Pacific', na=False)
df_aus_departures = df_departure_region[mask_pac]
df_aus_departures['delay in departure'].mean()

In [ ]:
mask_eu = df_departure_region['departure region'].str.contains('Europe', na=False)
df_aus_departures = df_departure_region[mask_eu]
df_aus_departures['delay in departure'].mean()

In [ ]:
mask_ame = df_departure_region['departure region'].str.contains('America', na=False)
df_aus_departures = df_departure_region[mask_ame]
df_aus_departures['delay in departure'].mean()

In [ ]:
mask_afr = df_departure_region['departure region'].str.contains('Africa', na=False)
df_aus_departures = df_departure_region[mask_afr]
df_aus_departures['delay in departure'].mean()

ASIA: 18.790226460071512  
AUS: 15.44949494949495  
Pacific: 14.6640625  
Europe: 17.633986928104576  
America: 15.333333333333334  
Africa: 14.318181818181818  

2.2 arrival delays

In [ ]:
df_arrival_region =df[['delay in arrival',  'arrival region']]
df_arrival_region

In [ ]:
mask_asia = df_arrival_region['arrival region'].str.contains('Asia', na=False)
df_asia_arrival = df_arrival_region[mask_asia]
df_asia_arrival['delay in arrival'].mean()

In [ ]:
mask_aus = df_arrival_region['arrival region'].str.contains('Australia', na=False)
df_aus_arrival = df_arrival_region[mask_aus]
df_aus_arrival['delay in arrival'].mean()

In [ ]:
mask_pac = df_arrival_region['arrival region'].str.contains('Pacific', na=False)
df_aus_arrival = df_arrival_region[mask_pac]
df_aus_arrival['delay in arrival'].mean()

In [ ]:
mask_eu = df_arrival_region['arrival region'].str.contains('Europe', na=False)
df_aus_arrival = df_arrival_region[mask_eu]
df_aus_arrival['delay in arrival'].mean()

In [ ]:
mask_ame = df_arrival_region['arrival region'].str.contains('America', na=False)
df_aus_arrival = df_arrival_region[mask_ame]
df_aus_arrival['delay in arrival'].mean()

In [ ]:
mask_afr = df_arrival_region['arrival region'].str.contains('Africa', na=False)
df_aus_arrival = df_arrival_region[mask_afr]
df_aus_arrival['delay in arrival'].mean()

ASIA: 6.968871595330739  
AUS: 6.709090909090909  
Pacific: 5.518518518518518  
Europe: 6.3125  
America: 7.5  
Africa: 3.8666666666666667  

### 3: delay with companies

In [114]:
df_airway_names = pd.read_csv('airway_names.csv', header=None, names=['airline'])

frequency_counts = df_airway_names['airline'].value_counts()
print(frequency_counts.head(20))


airline
'Qantas'                     62
'China Eastern Airlines'     55
'China Southern Airlines'    52
'Singapore Airlines'         45
'Air New Zealand'            42
'Korean Air'                 30
'American Airlines'          30
'ANA'                        29
'Xiamen Airlines'            26
'Delta Air Lines'            26
'Shenzhen Airlines'          24
'Emirates'                   24
'Qatar Airways'              23
'Air China LTD'              23
'VietJet Air'                21
'Etihad Airways'             20
'Azul'                       20
'KLM'                        20
'Shandong Airlines'          19
'Cathay Pacific'             19
Name: count, dtype: int64


In [127]:
df_company_names =df[['delay in departure', 'delay in arrival', 'airline names']]
df_company_names

delay in departure  delay in arrival          airline names
0                    NaN               NaN       'KLM Cityhopper'
1                    NaN              15.0           'CareFlight'
2                    8.0               NaN       'Cathay Pacific'
3                    5.0               NaN                'Scoot'
4                   14.0               NaN  'Philippine Airlines'
...                  ...               ...                    ...
1594                20.0               NaN   'Singapore Airlines'
1595                15.0               NaN   'Singapore Airlines'
1596                15.0               NaN              'EVA Air'
1597                 7.0               NaN      'Air New Zealand'
1598                 1.0               NaN                'SWISS'

[1599 rows x 3 columns]

Change the xxx of ['airline names'] == "'xxx'" to real airline names

In [160]:
filtered_df = df_company_names[df_company_names['airline names'] == "'Delta Air Lines'"]


In [161]:
filtered_df['delay in departure'].mean()

14.52

In [162]:
filtered_df['delay in arrival'].mean()

8.5

company departure arrival  
Qantas: 19.14516129032258, 6.478260869565218  
Air New Zealand: 15.463414634146341 7.0  
China Eastern Airlines: 15.132075471698114 6.714285714285714  
China Southern Airlines: 14.576923076923077 4.916666666666667  
Korean Air: 21.633333333333333 4.166666666666667  
Singapore Airlines: 20.829268292682926 7.4375  
Delta Air Lines: 14.52 8.5 
Qatar Airways: 20.695652173913043 8.428571428571429  
Emirates: 23.75 5.6  
British Airways: 11.583333333333334 12.0  

### 3: Delay to airports

In [167]:
df_airports = pd.read_csv('airport_names.csv', header=None, names=['departure airport names'])

frequency_counts = df_airports['departure airport names'].value_counts()
print(frequency_counts.head(20))

departure airport names
'airport': 'Singapore Changi                             57
'airport': 'Seoul (Incheon)                              53
'airport': 'Kuala Lumpur International Airport (klia)    44
'airport': 'Hong Kong International                      39
'airport': 'Perth International                          37
'airport': 'Melbourne - Tullamarine Airport              35
'airport': 'Suvarnabhumi International                   33
'airport': 'Ninoy Aquino International                   29
'airport': 'Tan Son Nhat International                   28
'airport': 'Narita International Airport                 28
'airport': 'Sydney Kingsford Smith Airport               25
'airport': 'Brisbane International                       25
'airport': 'Gimhae                                       24
'airport': 'Noibai International                         22
'airport': 'Kansai International                         21
'airport': 'Auckland International                       20
'airport': 'Haik

In [168]:
df_airports_dep = df[['delay in departure', 'departure airport names']]
df_airports_dep

delay in departure                            departure airport names
0                    NaN                               'airport': 'Oenpelli
1                    NaN                       'airport': 'Sydney Bankstown
2                    8.0                         'airport': 'Haneda Airport
3                    5.0  'airport': 'Taiwan Taoyuan International (Chia...
4                   14.0             'airport': 'Ninoy Aquino International
...                  ...                                                ...
1594                20.0                       'airport': 'Singapore Changi
1595                15.0                 'airport': 'Auckland International
1596                15.0                 'airport': 'Auckland International
1597                 7.0        'airport': 'Melbourne - Tullamarine Airport
1598                 1.0        'airport': 'Melbourne - Tullamarine Airport

[1599 rows x 2 columns]

In [210]:
filtered_df2 = df_airports_dep[df_airports_dep['departure airport names'] == "'airport': 'Shanghai Pudong International"]

In [211]:
filtered_df2['delay in departure'].mean()

17.38095238095238

Singapore Changi: 29.80821917808219  
Seoul (Incheon): 20.09090909090909  
Kuala Lumpur International Airport (klia): 19.75  
Hong Kong International: 22.2  
Perth International: 17.0  
Melbourne - Tullamarine Airport: 15.221052631578948  
Suvarnabhumi International: 31.363636363636363  
Auckland International: 14.84375  
Heathrow: 19.375  
Shanghai Pudong International: 17.38095238095238  


In [213]:
df_airports_arr = df[['delay in arrival', 'arrival airport names']]
df_airports_arr

delay in arrival                    arrival airport names
0                  NaN                   'airport': 'Maningrida
1                 15.0                  'airport': 'Williamtown
2                  NaN      'airport': 'Hong Kong International
3                  NaN             'airport': 'Singapore Changi
4                  NaN                        'airport': 'Malay
...                ...                                      ...
1594               NaN  'airport': 'Indira Gandhi International
1595               NaN                   'airport': 'Queenstown
1596               NaN   'airport': 'Christchurch International
1597               NaN       'airport': 'Brisbane International
1598               NaN             'airport': 'Singapore Changi

[1599 rows x 2 columns]

In [248]:
filtered_df3 = df_airports_arr[df_airports_arr['arrival airport names'] == "'airport': 'Shanghai Pudong International"]

In [249]:
filtered_df3['delay in arrival'].mean()

6.75

Singapore Changi: 7.407407407407407  
Seoul (Incheon): 5.25  
Kuala Lumpur International Airport (klia): 9.11111111111111  
Hong Kong International: 8.833333333333334  
Perth International: 5.545454545454546  
Melbourne - Tullamarine Airport: 6.625  
Suvarnabhumi International: 4.333333333333333  
Auckland International: 4.076923076923077  
Heathrow: 7.22222222222  
Shanghai Pudong International: 6.75  